cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -115,29 +120,41 @@
     this.programStateWriter = programStateWriter;
     this.tasks = new LinkedList<>();
     this.metricsCollectionService = metricsCollectionService;
+    this.transactionRunner = transactionRunner;
   }
 
   @Nullable
   @Override
+  protected String loadMessageId(StructuredTableContext context) throws IOException {
+    return getAppMetadataStore(context).retrieveSubscriberState(getTopicId().getTopic(), "");
+  }
+
+  @Override
+  protected void storeMessageId(StructuredTableContext context, String messageId) throws IOException {
+    getAppMetadataStore(context).persistSubscriberState(getTopicId().getTopic(), "", messageId);
+  }
+
+  @Nullable
+  @Override
+  // TODO: CDAP-14848 remove below two methods
```

Comment:
They are called, since some instances of the base class need to use dataset context, where other instances (such as this one) need to use structuredtablecontext. Thus, we declare both in the base class and the child classes implement one of the two. Once the migration is completed, we will delete the duplication

Reply:
ok